Add goal time (#4282)
dotnet/arcade: Documentation/Darc.md

- [delete-channel](#delete-channel)
- [get-channels](#get-channels)

> ### **`set-goal`**

Reviewer: I think "set-goal" is too broad.  I also wonder if having this command, in this fashion, kind of muddies the experience.  ie, if I run `> darc --help` and I see a list of commands, I kind of have an expectation about what those commands are doing and the general realm of things that I'm modifying.  I also, generally think that I'm changing / viewing something related to barviz or maestro api stuff.  

I'll caveat this by saying that I haven't actually dug into the darc code side of things, so I'm not sure if this is a reasonable suggestion or not.  Perhaps we should have this under like a "reporting" section which makes it clear that this is *different*.  ie `darc reporting set-goal-time`

Author: I can make the command name more descriptive. There are no sections as such in darc. This is going to be one of the commands will not trigger any visual changes in maestro/barViz. 

Probably make it something like this : 

darc set-goal-time-for-reporting : Sets an ideal build time for the definition. This is captured for the PKPI- Build time. Build time is calculated as a difference between start time of the official build till the time it takes for the build to be published to a channel.

Author: As discussed, I will look into adding new section in darc. I will keep you posted.

Author: I figured out a way to add reporting section into the command, I am trying this locally will update this thread again tomorrow.

Reviewer: To clarify further, if this just ends up being "to use this command, you have 2 random required parameters instead of 1 but otherwise the behavior is the same as any other command", then that's not really an interesting model.  My thought is something more like "git.exe" or "dotnet.exe" experience, where if you say "darc --help", you get some info like...

```
Microsoft.DotNet 1.1.0-beta.19454.3+1a295b33d669417b3d28795c3fa22efef21686a3
c Microsoft Corporation. All rights reserved.

  add-channel                Creates a new channel.

  add-dependency             Add a new dependency to version files.

  …

  reporting                          Additional commands used for PowerBI dashboard reporting
```

and then if write `darc reporting --help`, then you get info about the "set-goal-time" command.  If this is a lot of extra work, then it's probably not worth it and you can omit the whole "reporting" sub-command concept.

Author: As discussed since commandline lib does not have the ability to add this, so going back to the original plan on not adding section to this.

> Add or Update a goal time for existing repository for a specific channel. 

Reviewer 2: If a dev has no context, they'll likely wonder what "goal time" is for and how it's used.  I think it's probably smart to add a bit more - especially some prose regarding  what the goal is used for.  e.g. this is the goal time used for reporting our actual build times vs. goal....blah blah blah.  (or something)

Author: Sure will add text to this

> Add or Update a goal time for existing repository for a specific channel. 

Reviewer: nit: "Update" isn't  a command anymore, so it should be lower-case.

**Sample**:  
```

> PS D:\enlistments\arcade> darc set-goal  --goal 38 --repo "https://dev.azure.com/dnceng/internal/_git/dotnet-arcade" --channel ".Net Core 5 Dev" , ".Net Core 3.1 Release"

Reviewer: How does an end user decide whether to use the Azdo repo, or GitHub repo?  Does it matter which is used?

Author: Since we are using definitionId, repo is no longer relevant.

Added a new goal for Repo "https://dev.azure.com/dnceng/internal/_git/dotnet-arcade" = 38 minutes.
```

**Parameters**

> - `--channel` -  **(Required)**. Name of the channel/s where the repo is published.

Reviewer: should there be shortcuts for these other parameters just like there are for `--repo`?  ie `-c`?

Author: added shortcut for all the params

> - `-n, --name` -  **(Required)**. Name of the channel/s where the repo is published.

Reviewer 3: In the example you use --channel and in the parameter you use --name.  I am not sure which was intended but I like channel better

> - `-r, --repo` - **(Required)** Name of the repository.

Author: There is no good way to connect definition and goal time in BAR. Definition info resides in Kusto and repository info resides in SQL. 

I can connect them(timeline builds and goal details) after I get the data in Power BI. 

For the q, more than one Repo per definition, one will have to set goal time separately to both repos.

> - `-r, --repo` - **(Required)** Name of the repository.

Author: Regarding q,  Same Repo multiple definitions 

Will add a help text saying this will be applied to the official definitions only.

> - `-r, --repo` - **(Required)** Name of the repository.

Reviewer: It might simplify things a bit if this was definition id instead of repo.  It's a bit of a struggle to justify because it's merging in two very disjoint concepts and nothing else about darc has any clue about definitions (for good reason).  This entire concept of "goal lines" is a bolted on idea anyways, and maybe it's fine if it's unique in that way.  Given that definitions per channel is what is being shown, it's probably much simplier to tailor this command and corresponding table to your end goal than to try to fit it within the traditional darc / dependency flow model.

> - `-r, --repo` - **(Required)** Name of the repository.

Author: Spoke to Chris and Chad. We all agreed that we can use definitionId instead of repo. 

This is weird but I think we will live with definitionId residing in sql.

> - `--goal` - **(Required)** Goal time in minutues.

Reviewer: change this to `--minutes`?

- [delete-channel](#delete-channel)
- [get-channels](#get-channels)

### **`set-goal`**

> Sets a goal build time for the definition per Channel. This is captured for the PKPI- Build time report. Build time is calculated as a difference between start time of the official build till the time it takes for the build to be published to a channel. 

Reviewer 4: I don't think this clarification about the build time report is necessary here. This document serves as a user manual of sorts for darc, and as far as darc is concerned, you are just stating that you want to add some data about build time goals. How some other tooling reads that data is probably out of scope for this document.

Reviewer 2: My only comment here is that for most of the darc commands, it's implicitly understood "why" it's needed and how it relates to doing something.  This one is sorta just hanging out there, so a little more context might be in order.  In other words, a build time goal is only a meta-data thing....it has no build or dependency flow ramifications.

I don't super care either way here.  Do what you think best @epananth.  :)

**Sample**:  
```
PS D:\enlistments\arcade> darc reporting set-goal  --goal 38 --defintiionId 6  --channel ".Net Core 5 Dev" , ".Net Core 3.1 Release"

Added a new goal for definitionId 6 = 38 minutes.
```

**Parameters**

- `reporting` - **(Required)** Specifies that it's for PKPI- Build time purpose only. This is not for any visualization in BarViz or Maestro.

> - `-c,--channel` -  **(Required)** Name of the channel/s.

Reviewer 4: does this really take multiple channels?

Author: Yes

> - `-c,--channel` -  **(Required)** Name of the channel/s.

Reviewer 4: ```suggestion
- `-c, --channel` -  **(Required)** Name of the channel/s.
```

> - `-d, --definitionId` - **(Required)** Definition Id.

Reviewer 4: Do we need more information to disambiguate the definition id?

Are definition ids repeated among azdo orgs / projects ?

Reviewer 5: They are unique per... "project collection" I think is the word Azure DevOps uses?  So "dev.azure.com/dnceng", they are unique in that scope.

If there are builds we care about that live in some other project collection (probably "devdiv"), then this is insufficient.  But we don't monitor those today (they aren't pulled into our Telemetry tables), so it's not (currently) an issue.

Reviewer 4: Yeah, the specific example I was thinking about was the devdiv repos. I'd call it out explicitly in the document that this will only capture dnceng If we don't care about goal times for Roslyn / MSbuild /F# et al.

Author: As discussed, this is for dnceng only for now. Will add a comment in the doc to ensure that it is communicated.

Reviewer 6: Can you make the definition id more explicitly defined as the id that azure devops has assigned to the pipeline definition? I feel like just defining it as "Definition Id" isn't clear enough of where the id comes from.

> - `-m,--minutes` - **(Required)** Goal time in minutues.

Reviewer 4: I think you're still making changes to how the command will work, but there's a discrepancy between the sample and the parameters regarding the `--goal` and `--minutes` parameters.

> - `-m,--minutes` - **(Required)** Goal time in minutues.

Reviewer 4: ```suggestion
- `-m, --minutes` - **(Required)** Goal time in minutues.
```

Reviewer 6: You should take this fix.

- [delete-channel](#delete-channel)
- [get-channels](#get-channels)

### **`set-goal`**

> Sets a goal build time for the definition per Channel in minutes. Only dnceng is in scope. This is captured for the [Power BI dashboard](https://dev.azure.com/dnceng/public/_dashboards/dashboard/40ac4990-3498-4b3a-85dd-2ffde961d672). 

Reviewer 6: ```suggestion
Sets a goal build time for the definition per channel in minutes. Only dnceng is in scope. This is captured for the [Power BI dashboard](https://dev.azure.com/dnceng/public/_dashboards/dashboard/40ac4990-3498-4b3a-85dd-2ffde961d672). 
```

**Sample**:
```
PS D:\enlistments\arcade> darc set-goal --minutes 38 --definition-id 6 --channel ".Net Core 5 Dev"

38
```
**Parameters**

> - `-c,--channel` - **(Required)** Name of the channel.

Reviewer 6: ```suggestion
- `-c, --channel` - **(Required)** Name of the channel.
```

- `-d, --definition-id` - **(Required)** Azure DevOps Definition Id.
- `-m, --minutes` - **(Required)** Goal time in minutes.

### **`get-goal`**

Get the goal build time for a definition per channel in minutes. Only dnceng is in scope. This is captured for the [Power BI dashboard](https://dev.azure.com/dnceng/public/_dashboards/dashboard/40ac4990-3498-4b3a-85dd-2ffde961d672)

**Sample**:
```
PS D:\enlistments\arcade> darc get-goal --definition-id 6 --channel ".Net Core 5 Dev"

38
```
**Parameters**

> - `-c,--channel` - **(Required)** Name of the channel.

Reviewer 6: ```suggestion
- `-c, --channel` - **(Required)** Name of the channel.
```

- [delete-channel](#delete-channel)
- [get-channels](#get-channels)

### **`add-goal`**

Add a goal time for existing repository. The newly added goal time can be updated using [update-goal](#update-goal). 

**Sample**:  
```
( this is when we have a separate goal table in BAR )

PS D:\enlistments\arcade> darc add-goal  --goal 38 --repo "https://dev.azure.com/dnceng/internal/_git/dotnet-arcade" --channel ".Net Core 5 Dev"

Added a new goal for Repo "https://dev.azure.com/dnceng/internal/_git/dotnet-arcade" = 38 minutes.
```

> (Or)

Author: If I add this as a separate table it seems cleaner. But if I go with subscription table then we will have to supply only the subscriptionid other parameters can be optional.

- [delete-channel](#delete-channel)
- [get-channels](#get-channels)

### **`set-goal`**

> Sets a goal build time for the definition per Channel. This is captured for the PKPI- Build time report.

Reviewer 6: ```suggestion
Sets a goal build time for the definition per channel. This is captured for the PKPI - Build time report.
```

Reviewer: same suggestion as below (with link to dashboard)

**Sample**:  
```

> PS D:\enlistments\arcade> darc set-goal  --minutes 38 --defintiionId 6  --channel ".Net Core 5 Dev"

Reviewer 6: ```suggestion
PS D:\enlistments\arcade> darc set-goal  --minutes 38 --definitionId 6  --channel ".Net Core 5 Dev"
```

> PS D:\enlistments\arcade> darc set-goal  --minutes 38 --definitionId 6  --channel ".Net Core 5 Dev"

Reviewer: nit: there appear to be double spaces in this sample for now reason.

Added a new goal for definitionId 6 = 38 minutes.
```

**Parameters**

> - `-c,--channel` -  **(Required)** Name of the channel.

Reviewer 6: ```suggestion
- `-c, --channel` - **(Required)** Name of the channel.
```

> - `-d, --definitionId` - **(Required)** Definition Id.

Reviewer: nit: for consistency, this should probably be `--definition-id`

> - `-m, --minutes` - **(Required)** Goal time in minutues.

Reviewer: ```suggestion
- `-m, --minutes` - **(Required)** Goal time in minutes.
```

> 

Reviewer: nit: extra line

### **`get-goal`**

> Gets the goal build time for the definition per Channel. This is captured for the PKPI- Build time report.

Reviewer 6: ```suggestion
Gets the goal build time for the definition per channel. This is captured for the PKPI - Build time report.
```

> Gets the goal build time for the definition per Channel. This is captured for the PKPI- Build time report.

Reviewer: ```suggestion
Get the goal build time for a definition per channel in minutes. This is captured for the [Power BI dashboard](https://dev.azure.com/dnceng/public/_dashboards/dashboard/40ac4990-3498-4b3a-85dd-2ffde961d672)
```

**Sample**:  
```

> PS D:\enlistments\arcade> darc get-goal --defintiionId 6  --channel ".Net Core 5 Dev"

Reviewer 6: ```suggestion
PS D:\enlistments\arcade> darc get-goal --definitionId 6  --channel ".Net Core 5 Dev"
```

Reviewer: nit: spaces

> Goal for definitionId 6 = 38 minutes.

Reviewer: This could just print out `38`.  the other info is captured in the query line above and then the output would be more parseable (if it ever needs to be automated).

```

**Parameters**

> - `-c,--channel` -  **(Required)** Name of the channel.

Reviewer: nit: spacing

- `-d, --definitionId` - **(Required)** Definition Id.

> 

Reviewer: nit: extra line

**Sample**:  
```
PS D:\enlistments\arcade> darc set-goal  --minutes 38 --defintiionId 6  --channel ".Net Core 5 Dev"

Added a new goal for definitionId 6 = 38 minutes.
```

**Parameters**

- `-c,--channel` -  **(Required)** Name of the channel.
- `-d, --definitionId` - **(Required)** Definition Id.
- `-m,--minutes` - **(Required)** Goal time in minutues.

### **`get-goal`**

Gets the goal build time for the definition per Channel. This is captured for the PKPI- Build time report.

**Sample**:  
```
PS D:\enlistments\arcade> darc get-goal --defintiionId 6  --channel ".Net Core 5 Dev"

Goal for definitionId 6 = 38 minutes.
```

**Parameters**

- `-c,--channel` -  **(Required)** Name of the channel.
- `-d, --definitionId` - **(Required)** Definition Id.
- `-m,--minutes` - **(Required)** Goal time in minutues.

> 

Author: Keeping the params consistent for both set and get.

- [get-channels](#get-channels)
- [get-default-channels](#get-default-channels)
- [delete-default-channel](#delete-default-channel)

> - [set-goal](#set-goal)

Author: Spoke to @riarenas , @chcosta , @markwilkie and @Chrisboh 

adding another command 

get-goals - this will list the goal in minutes for the specific definition Id. 

Why are we adding this? 

- Right now there is no way to validate goal table other than the power bi report. Though this data is only used in power bi, we will still need to validate if the goals are being set or no. In case the report breaks then there is no good way to test this functionality. Where as if we have an extra command to check this in darc then we can eliminate the chances of failure for setting goal time in darc, so one can just focus on the fixing the power bi report.

Reviewer 2: Given that the only way to see the goals is via the report, is there a reason that this same report is not sufficient to see what the current goals are?

Author: We set the goal in darc, then we have to go the report to validate the changes. Sure I can add the link for the report and ask the user to test his/her changes in the report. But since this is a functionality in darc this has to be self sufficient and should not be dependent on an external tool for the validation of its data. That way we will decouple these two functionalities. 

Also, What if the report breaks? Then we cannot validate the changes in darc. We have to use Sql to check the data and find out if darc is failing or power bi is. This way it will make it easier to debug in case of errors, so maintaining this tool is going to be easier.

Reviewer 2: Thanks @epananth

Reviewer 5: I'm... skeptical this will ever be called. PowerBI doesn't often just "break", and the data it's pulling is pretty easy to see in the editor, so you can see what the value from SQL was. And it's pretty easy to log into the SQL DB on the rare occasion the data isn't present.  It's not going to be cheap to implement this command, because my understanding is that we'll need to write, deploy, and maintain an API that returns this value (since darc.exe doesn't have direct access to the DB).

I think it would be better to not implement this for now, and if we find the report breaking a lot and making the SQL query (which is just "SELECT * FROM Goals"), we can implement this later and pay the cost then.

Reviewer: Just to be clear, your recommend here is for no validation?

Reviewer 5: Personally? Yeah. The cost of developing and maintaining the API's in maestro is high, so unless there is a lot of value, I don't think it's worth it, for such a niche and low usage feature.

Author: Spoke to @ChadNedzlek and @chcosta 

The whole point for adding this extra change was to add validation for the command add-goal. Int API does not talk to Int SQL, it redirects to prod, so there is no way to validate this change unless we roll this out to prod. So we technically cannot validate this piece in staging/int. Adding another API to validate the changes will not help in this scenario. Also there is another issue in the backlog to add validation for all the API's.

Reviewer 7: Technically the only INT thing that redirects to prod is the API as accessed from the angular Web interface. If you use a token from INT and hit the INT API directly it will use the INT database.

Reviewer 5: Ooooo! That I didn't know. So "darc.exe" would actually be ok if we wanted to add this.  It's still a pain to add an API, but it's less burdensome now, so I'm far more "meh" about it now. :-) It might be worth doing if it's not too painful, and adding something to the "scenario-test.ps1" file... just a quick little "set-goal 50; get-goal == 50; set-goal 999; get-goal == 999"

Reviewer: @epananth , can you file an issue to follow up on this, but move forward without it for now?

Author: As discussed,

adding the command 

get-goal 

Why do we need to add this?

1) User will have the ability to check his/her work after he sets the goal.
2) For user experience, in command line if we have ability to set goal, we should also have the ability to get goal. Like that is pattern we are following for all the darc commands. 
3) The audience we are catering here is 4 to 5 folks for both the commands. Even though it is limited, I think we should give the best experience to our customers. 
4) Also how many of them will use the power bi report to see if the changes went through? This might come back to us saying they set the goal but power bi not updating.  So we will investigate this again in FR. Again this is a maintenance overhead.

- [get-channels](#get-channels)
- [get-default-channels](#get-default-channels)
- [delete-default-channel](#delete-default-channel)

> - [set-goal](#set-goal)

Reviewer: add get-goal section

- [add-default-channel](#add-default-channel)
- [delete-default-channel](#delete-default-channel)
- [get-default-channels](#get-default-channels)

> - [set-goal](#set-goal)

Reviewer: add get-goal section

**See also**:
- [get-subscriptions](#get-subscriptions)

### **`update-goal`**

Updated the goal time for existing repository.

**Sample**:
```
(Using new table to ingest goal)

PS D:\enlistments\arcade> darc update-goal  --goal 38 --repo "https://dev.azure.com/dnceng/internal/_git/dotnet-arcade" --channel ".Net Core 5 Dev"

Updated the goal for Repo "https://dev.azure.com/dnceng/internal/_git/dotnet-arcade" = 38 minutes.
```

> (OR)

Author: For the q,How to define a goal line for a repo that has published to multiple channels.

1) Can add the functionality to add channels as a list.

Reviewer 5: Can we just make it "set-goal" and it either adds or updates? It seems like there isn't much difference, so there isn't a real need to have 2 commands that have the same arguments and reach the same end state. :-)

Author: Sure I can do that. I thought about it. Since most of the darc commands have add and update, I tried to keep it in sync.

```
(if we add another column for subscription table)

PS D:\enlistments\arcade> darc update-goal  --goal 38 --repo "https://dev.azure.com/dnceng/internal/_git/dotnet-arcade" --channel ".Net Core 5 Dev" --subscriptionId "1abbb4c1-19d8-4912-fab8-08d6a19aff9"

Added a new goal for Repo "https://dev.azure.com/dnceng/internal/_git/dotnet-arcade" = 38 minutes.
```

**Parameters**

- `-n, --name` -  **(Required)**. Name of channel.
- `-r, --repo` - **(Required)** Name of the repository.
- `--goal` - **(Required)** Goal time in minutues.
- `--subscriptionid` - **(Required)** Subscription id. 

**See also**:
- [add-goal](#add-goal)

> 

Author: What happens when a new channel is added?
